### PR TITLE
WSE spec updated to include upstream and downstream sequencing

### PR DESCRIPTION
**Pull request by pkhanal**
Pending - 
* Update the robot tests
* Add robot tests

@jfallows, I Will update the robot tests and add new tests once the spec is finalized.

### CODE REVIEW

**Comment by pkhanal** [comment]
@jfallows, please review the changes.

Do we need a separate section to elaborate the upstream and downstream sequencing strategy?

**Comment by jfallows** [comment]
Looks great, @pkhanal !

Recommend burning the first `X-Sequence-No` on the create and use and incremented sequence number for the first upstream and downstream request instead.

Let's also make sure to specify the strict behavior when `X-Sequence-No` header is missing from the create request too.

Now we need to modify the existing scripts to use `X-Sequence-No` header and add another test for browser considerations to cover `.ksn` query parameter, and a few more negative tests for missing or out-of-order sequence number on both upstream and downstream.

File: SPEC.md [hunk 1]
```diff
@@ -137,9 +137,11 @@ To establish an emulated WebSocket connection, a client makes an HTTP handshake
   * `/;e/ctm` for mixed text encoding  (see [Binary as Mixed Text](#binary-as-mixed-text))
 * the HTTP handshake request path query parameters MUST include all query parameters from the WebSocket URL
 
-Browser clients MUST sent the `Origin` HTTP header with the source origin.
+Browser clients MUST send the `Origin` HTTP header with the source origin.
 
-Clients MUST sent the `X-WebSocket-Version` HTTP header with the value `wseb-1.1`.
+Clients MUST send the `X-WebSocket-Version` HTTP header with the value `wseb-1.1`.
+
+Clients MUST send the `X-Sequence-No` HTTP header with the sequence number that will serve as a starting value for upstream and downstream sequencing during data transfer. The sequence number MUST be a `positive integer` and cannot exceed 2^53 -1. It is recommended to choose the starting value such that the sequence number does not exceed the maximum value during data transfer. The sequence number MUST be maintained independently for upstream and downstream.
```
Comment (jfallows) [inline SPEC.md]: Strictly speaking, the `X-Sequence-No` sent by create does _not_ actually serve as the starting value for upstream and downstream sequencing.  Perhaps it would be useful to reword this to match the updated semantics where every request has a sequence number, even create, and then upstream and downstream diverge to increment independently.

File: SPEC.md [hunk 2]
```diff
@@ -240,7 +243,9 @@ If the server determines that any of the following conditions are not met by the
 send an HTTP response with a `4xx` status code, such as `400 Bad Request`.
 
 * the HTTP handshake request method MUST be `POST` 
-* the HTTP handshake request header `X-WebSocket-Version` MUST have the value `wseb-1.1`
+* the HTTP handshake request header `X-WebSocket-Version` MUST have the value `wseb-1.1`
+* the HTTP handshake request header `X-Sequence-No` is missing
```
Comment (jfallows) [inline SPEC.md]: This is a list of the positive conditions that must be met, otherwise the server sends a 404.  Therefore, since `X-Sequence-No` is required, this line should not be listed, right?

File: SPEC.md [hunk 3]
```diff
@@ -287,7 +292,8 @@ original handshake request URL path.
 Once the emulated WebSocket connection is established, the client MUST send an HTTP request for downstream data
 transfer.
 * the HTTP downstream request method MUST be `GET` 
-* the HTTP downstream request `Origin` header MUST be present with the source origin for browser clients
+* the HTTP downstream request `Origin` header MUST be present with the source origin for browser clients
+* the client MUST annotate the downstream request with sequence number. This is done via `X-Sequence-No` header which cannot exceed 2^53-1. The starting value of the header for downstream request MUST be increment of 1 from the value specified during [Opening Handshake](#opening-handshake)
```
Comment (jfallows) [inline SPEC.md]: We should say something about sequence numbers for all downstream requests not just the first one.

File: SPEC.md [hunk 4]
```diff
@@ -339,6 +346,20 @@ See [Binary as Mixed Text](#binary-as-mixed-text) for details of processing a mi
...
+The sequence number is regarded invalid - 
+
+* if it is not a positive integer
+* if the value exceeds 2^53 - 1
+
+The downstream request is regarded out of order - 
+
+* if the sequence number of first downstream request is NOT an increment of 1 compared to the value specified during [Opening Handshake](#opening-handshake)
+* if the sequence number of a downstream request is NOT an increment of 1 compared to that of preceding downstream request
+
```
Comment (jfallows) [inline SPEC.md]: Perhaps simplify to state that the downstream sequence number is out of order if it is not one greater than the previous downstream or create request?

File: SPEC.md [hunk 5]
```diff
@@ -359,13 +380,15 @@ attaching the downstream.
 
 Any upstream data frames are sent in the payload of a transient HTTP upstream request.
 * the HTTP upstream request method MUST be `POST`
-* the HTTP upstream request `Content-Type` HTTP header MUST be `application/octet-stream`
+* the HTTP upstream request `Content-Type` HTTP header MUST be `application/octet-stream`
+* the client MUST annotate the upstream request with sequence number. This is done via `X-Sequence-No` header which cannot exceed 2^53-1. The starting value of the header for upstream request MUST be increment of 1 from the value specified during [Opening Handshake](#opening-handshake)
```
Comment (jfallows) [inline SPEC.md]: Same feedback as downstream, semantics have changed, now one greater than value specified during create request.

File: SPEC.md [hunk 6]
```diff
@@ -395,6 +418,20 @@ with a `404 Not Found` status code.
...
+The sequence number is regarded invalid - 
+
+* if it is not a positive integer
+* if the value exceeds 2^53 - 1
+
+The upstream request is regarded out of order - 
+
+* if the sequence number of first upstream request is NOT an increment of 1 compared to the value specified during [Opening Handshake](#opening-handshake)
+* if the sequence number of a upstream request is NOT an increment of 1 compared to that of preceding upstream request
+
```
Comment (jfallows) [inline SPEC.md]: Simplify as per downstream feedback above?

File: src/main/java/org/kaazing/specification/wse/Functions.java [hunk 7]
```diff
@@ -57,6 +56,24 @@
...
+        int currentSequenceNo = Integer.valueOf(new String(current));
+        return String.valueOf(currentSequenceNo + 1).getBytes();
+    }
+
+    @Function
+    public static byte[] increment(byte[] current, int by) {
+        int currentSequenceNo = Integer.valueOf(new String(current));
+        return String.valueOf(currentSequenceNo + by).getBytes();
+    }
+
```
Comment (jfallows) [inline src/main/java/org/kaazing/specification/wse/Functions.java]: Try passing String instead of byte[] if int doesn't work.

File: SPEC.md (hunk 2 above)
Comment (pkhanal) [inline SPEC.md]: perhaps we can say - the HTTP handshake request header `X-Sequence-No` MUST be present ?

File: SPEC.md [hunk 8]
```diff
@@ -742,7 +759,19 @@ However, the client MAY choose not to supply the `.kb` parameter. Instead, the c
 when it decides it needs to (based on how much data has already been buffered on the current HTTP downstream response, and the 
 network connect latency). When the server detects the new downstream HTTP request, it will send a `RECONNECT` command on the 
 current downstream HTTP response, then complete it normally, and write all further data messages to the new downstream 
-HTTP response. 
+HTTP response.
+
+## Request Sequencing
+Every request (Create, Upstream & Downstream) in WSE connection MUST be annotated with sequence number. This is done via `X-Sequence-No` header. For platform which does not provide API support to add custom header to the HTTP request, sequencing can be achieved by using the query parameter `.kns`. The sequence number must be a `positive integer` and it cannot exceed 2 ^ 53 - 1. The sequence number of the subsequent request MUST be the increment of 1. The sequence number for Upstream and Downstream request diverge to increment independently once the WSE connection is established.
```
Comment (jfallows) [inline SPEC.md]: for platforms which do not provide API support to add custom headers

query parameter `.ksn` (not `.kns`)

"The sequence number of the subsequent request MUST be the increment of 1." - perhaps rephrase this sentence for clarity?

sequence number for Upstream and Downstream requests diverge (requests plural)

File: SPEC.md [hunk 9]
```diff
@@ -742,7 +759,19 @@ However, the client MAY choose not to supply the `.kb` parameter. Instead, the c
...
+
+A sequence number is regarded as invalid - 
+
+* if it is not a positive integer
+* if the value exceeds 2 ^ 53 - 1
```
Comment (jfallows) [inline SPEC.md]: Shouldn't we define valid and then state that all other forms are deemed invalid?

File: SPEC.md [hunk 10]
```diff
@@ -742,7 +759,19 @@ However, the client MAY choose not to supply the `.kb` parameter. Instead, the c
...
+
+A request is regarded `Out of Order` if the sequence number of the request is not `one` greater than previous request.
```
Comment (jfallows) [inline SPEC.md]: Similarly, perhaps we should define 'In Order' and then deem all others as 'Out of Order' ?

File: src/main/scripts/org/kaazing/specification/wse/browsers/client.request.padded.response/request.rpt [hunk 11]
```diff
@@ -1,3 +1,4 @@
+property sequence ${wse:generateRandom(100)}
```
Comment (jfallows) [inline src/main/scripts/org/kaazing/specification/wse/browsers/client.request.padded.response/request.rpt]: Suggest `randomInt` instead of `generateRandom` for function name.

File: src/main/scripts/org/kaazing/specification/wse/downstream/binary/request.out.of.order/downstream.response.rpt [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+# Downstream
+accept http://localhost:8080/path/T4hpc2vqXueYRyyl3dIQw5K64XLQxuf1
+accepted
+connected
+
+read await CREATED
+read method "GET"
+read version "HTTP/1.1"
+read header "X-Sequence-No" ${wse:asString(sequence + 2)}
```
Comment (jfallows) [inline src/main/scripts/org/kaazing/specification/wse/downstream/binary/request.out.of.order/downstream.response.rpt]: We should file an enhancement request for K3PO to support negative matches for header value that include an EL variable, such as sequence above.

This script verifies the correct behavior when faced with an out-of-order request of sequence + 2, but should really be verifying the correct behavior when faced with an out-of-order request of anything except sequence + 1.

File: src/main/scripts/org/kaazing/specification/wse/opening/request.header.x.websocket.protocol/handshake.response.rpt [hunk 13]
```diff
@@ -14,7 +16,6 @@ write header "X-WebSocket-Protocol" "primary"
 write header "Content-Type" "text/plain;charset=utf-8"
 write header content-length
 
-write "http://localhost:8080/path/uofdbnreiodfkbqi"
-write "\n"
-write "https://localhost:8443/path/kwebfbkjwehkdsfa"
+write "http://localhost:8080/path/uofdbnreiodfkbqi\n"
+write "https://localhost:8443/path/kwebfbkjwehkdsfa\n"
```
Comment (jfallows) [inline src/main/scripts/org/kaazing/specification/wse/opening/request.header.x.websocket.protocol/handshake.response.rpt]: Nice tidying. :smile:

**Comment by jfallows** [comment]
Excellent work, @pkhanal !  Mostly just minor feedback. :smile:

File: src/main/scripts/org/kaazing/specification/wse/downstream/binary/request.out.of.order/downstream.response.rpt (hunk 12 above)
Comment (pkhanal) [inline src/main/scripts/org/kaazing/specification/wse/downstream/binary/request.out.of.order/downstream.response.rpt]: Agree!

**Comment by pkhanal** [comment]
@jfallows, minor feedback has been applied. One step closer to merge :)

File: SPEC.md [hunk 14]
```diff
@@ -742,7 +759,16 @@ However, the client MAY choose not to supply the `.kb` parameter. Instead, the c
...
+
+A sequence number is **valid** if it is a `positive integer` AND it is less than or equal to `2 ^ 53 - 1`. Otherwise, the sequence number is **invalid**.
+
+A request is regarded `In Order` if the sequence number of the request is `one` greater than previous request. Otherwise, the request is treated as `Out of Order` request.
+
+If the sequence number of the Create request is 10, the sequence number of subsequent upstream and downstream request MUST be `11`. During data transfer, the sequence number of subsequent upstream and downstream request increment independently. The sequence number of a Downstream request MUST be `one` greater than the sequence number of the previous Create or Downstream request. Likewise, the sequence number of an Upstream request MUST be `one` greater than the sequence number of the previous Create or Upstream request.
```
Comment (jfallows) [inline SPEC.md]: For example...
